flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java

```diff
@@ -376,8 +377,14 @@ private static CompletedCheckpoint retrieveCompletedCheckpoint(Tuple2<Retrievabl
 
 		try {
 			return stateHandlePath.f0.retrieveState();
-		} catch (Exception e) {
-			throw new FlinkException("Could not retrieve checkpoint " + checkpointId + ". The state handle seems to be broken.", e);
+		} catch (ClassNotFoundException cnfe) {
+			throw new FlinkException("Could not retrieve checkpoint " + checkpointId + " from state handle under " +
+				stateHandlePath.f1 + ". This indicates that you are trying to recover from state written by an " +
+				"older Flink version which is not compatible. Try cleaning the state handle store.", cnfe);
+		} catch (IOException ioe) {
+			throw new FlinkException("Could not retrieve " + checkpointId + " worker from state handle under " +
```

Comment:
shouldn't this say `Could not retrieve checkpoint " + checkpointId + " from state handle under` like in case of an CNFE?